File: Code/disToRanks.m
function [disRank]=disToRanks(dis) %transform from distance to ranking, order from 0,...,n-1. 

n=size(dis,1);
[~,ind]=sort(dis(1:n,1:n));
disRank=zeros(n,n);
for i=1:n
    v=ind(1:n,i);
    for j=1:n
        disRank(v(j),i)=j-1;
    end
    if disRank(i,i)~=0 %change the diagonal rank to zero, if necessary
        disRank(v(1),i)=disRank(i,i);
        disRank(i,i)=0;
    end
end

File: Code/R/disToRanks.R
# source("disToRanks.R")

disToRanks <- function(dis) {
  # An auxiliary function that sorts the entries within each column by ascending order.
  #
  # The input is assumed to be a distance matrix.
  #
  # The output is column-wise rank, ordered from 1,...,n.
  #
  # For ties, the minimum ranking is used, 
  # e.g. if there are repeating distance entries, the order is like 1,2,3,3,4,..,n-1.
  n=nrow(dis);
  disRank=matrix(0,n,n);
  for (i in (1:n)){
    v=dis[,i];
    tmp=rank(v,ties.method="min");
    tu=unique(tmp);
    if (length(tu)!=max(tmp)){
      tu=sort(tu);
      for (j in (1:length(tu))){
        kk=which(tmp==tu[j]);
        tmp[kk]=j;
      }
    }
    disRank[,i]=tmp;
  }
  return(disRank);
}